cranelift-codegen/src/legalizer/mod.rs

```diff
@@ -665,3 +665,69 @@ fn narrow_iconst(
 
     unimplemented!("missing encoding or legalization for iconst.{:?}", ty);
 }
+
+fn narrow_icmp_imm(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    isa: &dyn TargetIsa,
+) {
+    use crate::ir::condcodes::{CondCode, IntCC};
+
+    let (arg, cond, imm): (ir::Value, IntCC, i64) = match func.dfg[inst] {
+        ir::InstructionData::IntCompareImm {
+            opcode: ir::Opcode::IcmpImm,
+            arg,
+            cond,
+            imm,
+        } => (arg, cond, imm.into()),
+        _ => panic!("unexpected instruction in narrow_icmp_imm"),
+    };
+
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    let ty = pos.func.dfg.ctrl_typevar(inst);
+    if isa.pointer_bits() == 32 && ty == I64 {
```

Comment:
Please drop this check and use `ty.half_width().unwrap()` instead of `I32`, `ty.half_width().unwrap().bits()` instead of 32 and `imm & (1u128 << ty.half_width().unwrap().bits() - 1) as i64` instead of `imm & 0xffffffff`. Legalizations have lower precedence than encodings, so this wouldn't get called for `i64` on 64bit x86 anyway. As already noted in https://github.com/CraneStation/cranelift/pull/1081#pullrequestreview-294285205, legalizing this for `i128` on both 32bit and 64bit x86 would be very useful for me.

Reply:
@bjorn3 on it :)

Reply:
@bjorn3 done! PTAL.